File: lib/precision.ts
// lib/precision.ts
//
// PRECISION CORE — truncate at 2 decimals, no rounding.
// 3.088 -> 3.08, 3.099 -> 3.09. toPrecision(15) scrubs binary
// float noise first so a value stored as 3.0799999999 does not
// wrongly truncate to 3.07. All intermediate math stays in full
// precision; truncation happens only at the display boundary.

/** Truncate a number to exactly 2 decimals (toward zero), drift-safe. */
export function trunc2(num: number): number {
  if (!Number.isFinite(num)) return NaN;
  const sign = num < 0 ? -1 : 1;
  const scaled = Number((Math.abs(num) * 100).toPrecision(15));
  return (Math.trunc(scaled) / 100) * sign;
}

/** Format a number as a fixed 2-decimal string via truncation. */
export function fmt2(num: number): string {
  const t = trunc2(num);
  return Number.isFinite(t) ? t.toFixed(2) : "0.00";
}

/** Add thousands separators to the integer part of a "1234.56" string. */
export function group(str: string): string {
  const [i, d] = str.split(".");
  const neg = i.startsWith("-");
  const digits = neg ? i.slice(1) : i;
  const grouped = digits.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  return (neg ? "-" : "") + grouped + (d !== undefined ? "." + d : "");
}

/** Format as naira with the ₦ symbol and 2-decimal truncation. */
export function naira(n: number): string {
  return "\u20A6" + group(fmt2(n));
}

/** Format as kg value with 2-decimal truncation. */
export function kg(n: number): string {
  return group(fmt2(n));
}

export interface Parsed {
  empty: boolean;
  value: number;
  nan: boolean;
}

/** Parse a raw input string, distinguishing empty from zero. */
export function parseField(raw: string | null | undefined): Parsed {
  const s = (raw ?? "").trim();
  if (s === "") return { empty: true, value: NaN, nan: false };
  const v = Number(s.replace(/,/g, ""));
  return { empty: false, value: v, nan: !Number.isFinite(v) };
}

File: android-app/app/src/main/java/com/gasrefill/terminal/utils/Precision.kt
package com.gasrefill.terminal.utils

import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

object Precision {
    private val format = DecimalFormat("#,###.00", DecimalFormatSymbols(Locale.US)).apply {
        isDecimalSeparatorAlwaysShown = true
        minimumFractionDigits = 2
        maximumFractionDigits = 2
        isGroupingUsed = true
    }

    /** Truncate to 2 decimal places (toward zero). */
    fun trunc2(num: Double): Double {
        if (!num.isFinite()) return Double.NaN
        val sign = if (num < 0) -1.0 else 1.0
        val scaled = kotlin.math.abs(num) * 100.0
        return (scaled.toLong().toDouble() / 100.0) * sign
    }

    /** Format with thousands separator and 2 decimals. */
    fun fmt2(num: Double): String {
        val t = trunc2(num)
        if (!t.isFinite()) return "0.00"
        return format.format(t)
    }

    /** Format as naira with symbol. */
    fun naira(num: Double): String = "\u20A6${fmt2(num)}"

    /** Format as kg value. */
    fun kg(num: Double): String = fmt2(num)

    /** Parse a raw input string. */
    data class Parsed(val empty: Boolean, val value: Double, val nan: Boolean)

    fun parseField(raw: String?): Parsed {
        val s = (raw ?: "").trim().replace(",", "")
        if (s.isEmpty()) return Parsed(true, Double.NaN, false)
        val v = s.toDoubleOrNull()
        return if (v != null) Parsed(false, v, false)
        else Parsed(false, Double.NaN, true)
    }
}
